Source/WebCore/loader/PingLoader.cpp

```diff
@@ -146,9 +146,15 @@ void PingLoader::sendPing(Frame& frame, const URL& pingURL, const URL& destinati
     FrameLoader::addHTTPOriginIfNeeded(request, SecurityPolicy::generateOriginHeader(document.referrerPolicy(), request.url(), sourceOrigin));
 
     frame.loader().updateRequestAndAddExtraFields(request, IsMainResource::No);
+
+    // https://www.w3.org/TR/2009/WD-html5-20090423/history.html#hyperlink-auditing
     request.setHTTPHeaderField(HTTPHeaderName::PingTo, destinationURL.string());
-    if (!SecurityPolicy::shouldHideReferrer(pingURL, frame.loader().outgoingReferrer()))
-        request.setHTTPHeaderField(HTTPHeaderName::PingFrom, document.url().string());
+    if (document.securityOrigin().isSameOriginAs(SecurityOrigin::create(pingURL).get()))
```

Comment:
Could we match the conditions in the spec more closely? I find it a lot less error prone:
- If the [URL](https://dom.spec.whatwg.org/#concept-document-url) of the [Document](https://html.spec.whatwg.org/multipage/dom.html#document) object containing the hyperlink being audited and ping URL have the [same origin](https://html.spec.whatwg.org/multipage/browsers.html#same-origin)
- If the origins are different, but the [scheme](https://url.spec.whatwg.org/#concept-url-scheme) of the [URL](https://dom.spec.whatwg.org/#concept-document-url) of the [Document](https://html.spec.whatwg.org/multipage/dom.html#document) containing the hyperlink being audited is not "https"
request must include a `[Ping-From](https://html.spec.whatwg.org/multipage/links.html#ping-from)` header with, as its value, the [URL](https://dom.spec.whatwg.org/#concept-document-url) of the document containing the hyperlink, and a `[Ping-To](https://html.spec.whatwg.org/multipage/links.html#ping-to)` HTTP header with, as its value, the target URL.
- Otherwise
request must include a `[Ping-To](https://html.spec.whatwg.org/multipage/links.html#ping-to)` HTTP header with, as its value, target URL. request does not include a `[Ping-From](https://html.spec.whatwg.org/multipage/links.html#ping-from)` header.

Reply:
I was following the logic of the old algorithm quite closely, but the new link has a new algorithm :|